tests/test_derived_cameras.py

```diff
@@ -0,0 +1,351 @@
+        """
+        Check that if warping is applied to a mask image (a few integer classes)
+        those classes are preserved.
+        """
+
+        image = np.ones((21, 21), dtype=np.uint8)
+        image[:10] = 0
+        image[:, 5:] = 2
+
+        cameras = MetashapeCameraSet(
+            camera_file=camera_file(tmp_path),
+            image_folder=tmp_path,
+        )
+        camera = simplify_camera(cameras.cameras[0], image)
+        camera.distortion_params["k1"] = k1
+
+        # Warp the image
+        dewarped = cameras.warp_dewarp_image(
+            camera, image, warped_to_ideal=w2i, inversion_downsample=downsample
+        )
+
+        # Check that we kept our desired classes (no statements about position)
+        assert sorted(np.unique(dewarped)) == [0, 1, 2]
+
+    @pytest.mark.parametrize(
+        "delete", combinations(["b1", "b2", "k1", "k2", "k3", "k4", "p1", "p2"], 2)
+    )
+    @pytest.mark.parametrize(
+        "w2i,k1,relationship",
+        (
+            [True, 1, operator.lt],
+            [True, -1, operator.gt],
+            [False, 1, operator.gt],
+            [False, -1, operator.lt],
+        ),
+    )
+    @pytest.mark.parametrize("downsample", [1, 2])
+    def test_dropped_parameters(
+        self, tmp_path, gradient, delete, w2i, k1, relationship, downsample
+    ):
+        """
+        Check that if distortion parameters are dropped we continue on. Arbitrarily
+        drop combinations of two parameters, skipping k1.
+        """
+
+        cameras = MetashapeCameraSet(
+            camera_file=camera_file(tmp_path),
+            image_folder=tmp_path,
+        )
+        camera = simplify_camera(cameras.cameras[0], gradient, delete=delete)
+        camera.distortion_params["k1"] = k1
+
+        # Warp the image
+        dewarped = cameras.warp_dewarp_image(
+            camera, gradient, warped_to_ideal=w2i, inversion_downsample=downsample
+        )
+
+        # Check whether the image on the average got lighter, darker, or
+        # stayed the same
+        assert dewarped.shape == gradient.shape
+        assert relationship(dewarped.mean(), gradient.mean())
+
+
+class TestPix2Face:
+
+    def test_simple_camera(self):
+        """Warp/dewarp should fail if not given a Metashape camera."""
+
+        # Create simple mesh and camera
+        mesh, point_colors = make_simple_mesh(pixels=[], color=None)
+        textured_mesh = TexturedPhotogrammetryMesh(
+            mesh=mesh,
+            input_CRS=EARTH_CENTERED_EARTH_FIXED_CRS,
+            texture=point_colors,
+        )
+        cameras = make_simple_camera_set()
+
+        # Render the mesh from these cameras
+        with pytest.raises(NotImplementedError):
+            textured_mesh.pix2face(
+                cameras=cameras,
+                cache_folder=None,
+                distortion_set=cameras,
+                apply_distortion=True,
+            )
+
+    @pytest.mark.parametrize("render_img_scale", [0.7, 0.8, 0.9, 1.0])
```

Comment:
Nice that it tests the scale here. In practice, we'd often use something like 0.5 or even 0.1. Maybe add that to the list?

Reply:
We can do 0.5, in this case the test image is low-res enough that going all the way to 0.1 means there are too few pixels to make good statements about. This won't make a difference in practice since the real images are ~150 times larger